src/Symfony/Bundle/FrameworkBundle/Client.php

```diff
@@ -104,6 +106,45 @@ public function enableReboot()
         $this->reboot = true;
     }
 
+    /**
+     * Logins a user against a specific firewall.
+     *
+     * @param string|object $user
+     * @param string        $credentials
+     * @param string[]      $roles
+     * @param string        $firewallName
+     * @param string|null   $firewallContext
+     */
+    public function login(
+        $user,
+        string $credentials,
+        array $roles = array('ROLE_ADMIN'),
+        string $firewallName = 'main',
+        string $firewallContext = null
+    ) {
+        if (!class_exists(UsernamePasswordToken::class)) {
+            throw new \RuntimeException('You must install the Symfony Security component to use this feature.');
+        }
+
+        $context = $firewallContext ?? $firewallName;
+        $session = $this->getContainer()->get('session');
+
+        $token = new UsernamePasswordToken($user, $credentials, $firewallName, $roles);
```

Comment:
I think creating the token should be put in a private method so we can override it, in case one needs to check different types of authentication modes that create different kinds of tokens.

Reply:
Sorry, I meant protected, I should be a bit tired :sweat_smile:

Reply:
but then, it should not be in the Client class either (which is not as easy to replace than a method in the TestCase).
And I'm not sure we really need to make it extendable, as everything it does here can be done from outside the client if you have different need.

Reply:
@Pierstoval I believe that if people need a more advanced way to login, they should not rely on this method. IMO, that's a reasonable limitation considering that it will ease most "login" use cases. WDYT ?